### ISSUE TITLE
Save results by script / create command history script lacks second Backslash & saved file lacks dot infront of .txt

### ISSUE THREAD

**Issue by DavidMHaumann** [qupath/qupath#125]
Dear all, 

i use the following line in the script editor to automatically save the results from annotation measurements: 

saveAnnotationMeasurements('C:\\Users\\d-hau\\Jobs on SSD\\Result Tables', );

The create command history script will show me the code with one backslash only: 
saveAnnotationMeasurements('C:\Users\d-hau\Jobs on SSD\Result Tables', );

I need to correct it manually each time to make it run. Is there a secret to fix that?

The second thing is: 
The result will be saved as .txt file. It is written with the name of the original file attached with txt. 

Example: 
Filename:   HESlide
Result: HESlide annotationstxt

instead of HE Slide annotations.txt

Only with dot, windows will recognise the file format. 

Is there a way to change the script to add the dot infront of "txt"?

Best
David

**Comment by DavidMHaumann** [comment]
and another thing: 
Is it possible to dynamicly create the target folder for the results?
If QuPath loads an old script into a new project, the pathway for the "Save results" part in the script need to be manually adapted to the place/folder of the new project. 

It would be nice, the script creates automatically a new folder "results" and saves the annotations results inside the new results folder in the new project. 

To create a new folder works like that: 
// Create the output directory, if required
def path = QPEx.buildFilePath(QPEx.PROJECT_BASE_DIR, "results")
QPEx.mkdirs(path)


But I dont know enought code yet to direct the new annotation results into the new folder.

**Comment by Svidro** [comment]
I am fairly certain that the Windows slash issue is listed as one of the things to fix already in another posting already, so might get done in a later version of QuPath.  If you are using the same location each time, you can just create a separate script and copy and paste the command in whenever you need it.  If this is for different projects and different directories... then yes, I think, for Windows, that is the only way to do it.

You can call a Python script on the folder where the file is placed to fix all of the .txt files, use a program like Bulk Rename Utility (I made quite a lot of use of that before learning any Python) or when you are already editing the path string, you could manually entire the file name as well.  If it is for a batch run, you will need to get the name of the image from something like:

outputFolder = "D:\\Results\\"
String imageLocation = getCurrentImageData().getServer().getPath()
fileNameWithNoExtension = imageLocation.split("[^A-Za-z0-9_ ]")[-2]
saveAnnotationMeasurements(outputFolder+fileNameWithNoExtension+".txt", )

I may be missing a few useful characters in the split string there, but I suspect you get the idea!

**Comment by Svidro** [comment]
For the second part, I am guessing you would want to define the above outputFolder using something like:

```
import qupath.lib.gui.QuPathGUI
def project = QuPathGUI.getInstance().getProject().getBaseDirectory()
println(project)
```

**Comment by DavidMHaumann** [comment]
Hi Svidro, 
thanks a lot for your ideas. 

For now, I work with this: 

import qupath.lib.scripting.QPEx
// Create the output directory, if required
def path = QPEx.buildFilePath(QPEx.PROJECT_BASE_DIR,"results")
QPEx.mkdirs(path)

String imageLocation = getCurrentImageData().getServer().getPath()
fileNameWithNoExtension = imageLocation.split("[^A-Za-z0-9_ ]")[-2]
saveAnnotationMeasurements(path+"_"+fileNameWithNoExtension+".txt", )


It will write the result .txt with the right ending into the project folder. 
Better than nothing. 
At the same time it will write the new project folder but not use it. I do not really get why. 
Instead it is using the name of the new folder als prefix of the result file!
Means "resultsNAMEofFile.txt"  is written in the project folder

**Comment by DavidMHaumann** [comment]
to find the project folder by definition of path, the single backslashes seem not to be a problem. 
But for the final result folder, it does! 
Strange.

**Comment by Svidro** [comment]
For the results part, you just need an extra set of //.  As you have it
now, you are writing out to projectfolder/results_imagename.txt because
path is projectfolder/results not projectfolder/results/.

You either need to add the / or // (double slash for Windows) on to the end
of the path, or at the beginning of the file name to use results at a
folder.

Handling the whole thing is a little weird since single slashes work just
fine when the path generated by the program is being used on its own, but
as soon as you treat the path as a string, the fact that it is created with
single / becomes a problem.

**Comment by petebankhead** [comment]
I found the earlier reference at #25 but this thread describes the problem better.  It may be worth a look at the final posts in that issue though, for an alternative workaround to the `.txt` problem.

Definitely a bug that needs fixing...

**Comment by DavidMHaumann** [comment]
Hi Pete, 
yes - I followed the instructions from Issue 25 and it worked out! Very nice! Thank you. 

Cheers :-)

**Comment by petebankhead** [comment]
This issue should be resolved in the latest milestone versions (v0.2.0-mx).
